src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala

```diff
@@ -0,0 +1,126 @@
+  def withFixture(eventStreamMaxOutstandingMessages: Int = 5)(fn: Fixtures => Unit): Unit = {
+    val f = new Fixtures(eventStreamMaxOutstandingMessages)
+    try (fn(f)) finally {
+      f.leaderStateEventsInput.complete()
+    }
+  }
+
+  class Fixtures(val eventStreamMaxOutstandingMessages: Int = 5) {
+    implicit val electionService = mock[ElectionService]
+    electionService.leaderHostPort returns Some("host:1")
+    electionService.localHostPort returns "host:1"
+    electionService.isLeader returns true
+    val (leaderStateEventsInput, leaderStateEvents) = Source.queue[LeadershipState](16, OverflowStrategy.fail)
+      .toMat(Repeater.sink(16, OverflowStrategy.fail))(Keep.both)
+      .run
+    electionService.leaderStateEvents returns leaderStateEvents
+    implicit val authorizer = mock[Authorizer]
+    implicit lazy val authenticator = mock[Authenticator]
+
+    authorizer.isAuthorized(any, any, any) returns true
+    authenticator.authenticate(any) returns Future.successful((Some(new Identity {})))
+
+    val eventBus = new EventStream(system)
+    val eventsController = new EventsController(
+      eventStreamMaxOutstandingMessages = eventStreamMaxOutstandingMessages,
+      eventBus = eventBus)
+
+    /**
+      * Makes a single request. We don't use the akka testkit `check` methods here because they all want to consume the
+      * entire response before making any data available.
+      *
+      * Read [[https://doc.akka.io/docs/akka-http/10.0.10/java/http/routing-dsl/testkit.html#writing-asserting-against-the-httpresponse]]
+      */
+    def eventsRequest(request: HttpRequest): HttpResponse = {
+      Source.single(request)
+        .via(Route.handlerFlow(eventsController.route))
+        .runWith(Sink.head)
+        .futureValue
+    }
+  }
+
+  "It publishes an event_stream_attached event upon connection for the connected IP" in withFixture() { f =>
+    val response = f.eventsRequest(Get().addHeader(`X-Real-Ip`(RemoteAddress(Array[Byte](1, 2, 3, 4)))))
+
+    val attachedEvent = response.entity.dataBytes
+      .via(EventStreamParser(Int.MaxValue, Int.MaxValue))
+      .runWith(Sink.head)
+      .futureValue
+
+    attachedEvent.`type` shouldBe Some("event_stream_attached")
+    attachedEvent.data.get should include("1.2.3.4")
+  }
+
+  "Relays events that are published after it is connected" in withFixture() { f =>
+    val response = f.eventsRequest(Get())
+
+    val output = response.entity.dataBytes
+      .via(EventStreamParser(Int.MaxValue, Int.MaxValue))
+      .drop(1)
+      .take(1)
+      .runWith(Sink.head)
+
+    f.eventBus.publish(testTerminatedEvent)
+
+    val sse = output.futureValue
+    sse.`type` shouldBe Some("app_terminated_event")
+    sse.data.get should include("/my-app")
+  }
+
+  "Respects filtering conditions" in withFixture() { f =>
+    val response = f.eventsRequest(Get(Uri("/?event_type=app_terminated_event")))
+
+    val output = response.entity.dataBytes
```

Comment:
What type is `output` here?

Reply:
output is the materialized type of the stream. Down below, on line 103, you see `.runWith(Sink.queue())`.

Looking at the docs for [.runWith](https://doc.akka.io/api/akka/current/akka/stream/scaladsl/Source.html#runWith[Mat2](sink:akka.stream.Graph[akka.stream.SinkShape[Out],Mat2])(implicitmaterializer:akka.stream.Materializer):Mat2), we see that it returns the materialized type for the provided Sink:

```
runWith[Mat2](sink: Graph[SinkShape[Out], Mat2])(implicit materializer: Materializer): Mat2
```

The sink we pass it is of type [Sink.queue](https://doc.akka.io/api/akka/current/akka/stream/scaladsl/Sink$.html#queue[T]():akka.stream.scaladsl.Sink[T,akka.stream.scaladsl.SinkQueueWithCancel[T]]), which has this type signature:

```
Sink[T, SinkQueueWithCancel[T]]
```

Therefore, the type of `output` is [SinkQueueWithCancel](https://doc.akka.io/api/akka/current/akka/stream/scaladsl/SinkQueueWithCancel.html)